src/common/cache.js

```diff
@@ -1,40 +1,61 @@
-const defaults = {
-  lifetime: 3000,
-};
-
-export default function initCache(options) {
-  const cache = {};
-  const { lifetime: defaultLifetime } = options || defaults;
+export default function initCache({
+  lifetime: defaultLifetime = 3000,
+} = {}) {
+  let cache = {};
+  // setTimeout call is very expensive when done frequently,
+  // 1000 calls performed for 50 scripts consume 50ms on each tab load,
+  // so we'll schedule trim() just once per event loop cycle,
+  // and then trim() will trim the cache and reschedule itself to the earliest expiry time.
+  let timer;
   return {
     get, put, del, has, hit, destroy,
   };
   function get(key, def) {
     const item = cache[key];
     return item ? item.value : def;
   }
-  function put(key, value, lifetime) {
-    del(key);
+  function put(key, value, lifetime = defaultLifetime) {
     if (value) {
       cache[key] = {
         value,
-        timer: setTimeout(del, lifetime || defaultLifetime, key),
+        expiry: lifetime + performance.now(),
       };
+      if (!timer) timer = setTimeout(trim);
+    } else {
+      delete cache[key];
     }
   }
   function del(key) {
-    const item = cache[key];
-    if (item) {
-      if (item.timer) clearTimeout(item.timer);
-      delete cache[key];
-    }
+    delete cache[key];
   }
   function has(key) {
-    return !!cache[key];
+    return Object.hasOwnProperty.call(cache, key);
   }
-  function hit(key, lifetime) {
-    put(key, get(key), lifetime);
+  function hit(key, lifetime = defaultLifetime) {
+    const entry = cache[key];
+    if (entry) {
+      entry.expiry = performance.now() + lifetime;
+      if (!timer) timer = setTimeout(trim);
+    }
   }
   function destroy() {
-    Object.keys(cache).forEach(del);
+    clearTimeout(timer);
+    timer = 0;
+    cache = {};
+  }
+  function trim() {
+    const now = performance.now();
+    let closest = Number.MAX_SAFE_INTEGER;
+    for (const key in cache) {
+      if (Object.hasOwnProperty.call(cache, key)) {
```

Comment:
How about this?
```js
for (const key of Object.keys(cache)) {
```

Reply:
Possible, if you insist, but usually I dislike the idea of temporary creating a big array just because there's a common prejudice in JS world against for-in...

Reply:
OK, we can just leave it.

Reply:
What if we switch to Map()?

Reply:
Like this:
```js
export default function initCache({
  lifetime: defaultLifetime = 3000,
} = {}) {
  let cache = new Map();
  // setTimeout call is very expensive when done frequently,
  // 1000 calls performed for 50 scripts consume 50ms on each tab load,
  // so we'll schedule trim() just once per event loop cycle,
  // and then trim() will trim the cache and reschedule itself to the earliest expiry time.
  let timer;
  let minLifetime = -1;
  return {
    get, put, del, has, hit, destroy,
  };
  function get(key, def) {
    const item = cache[key];
    return item ? item.value : def;
  }
  function put(key, value, lifetime = defaultLifetime) {
    if (value) {
      const expiry = lifetime + performance.now();
      cache.set(key, { value, expiry });
      reschedule(lifetime);
    } else {
      delete cache[key];
    }
  }
  function del(key) {
    cache.delete(key);
  }
  function has(key) {
    return Object.hasOwnProperty.call(cache, key);
  }
  function hit(key, lifetime = defaultLifetime) {
    const entry = cache.get(key);
    if (entry) {
      entry.expiry = performance.now() + lifetime;
      reschedule(lifetime);
    }
  }
  function destroy() {
    clearTimeout(timer);
    timer = 0;
    cache.clear();
  }
  function reschedule(lifetime) {
    if (timer) {
      if (lifetime >= minLifetime) return;
      clearTimeout(timer);
    }
    minLifetime = lifetime;
    timer = setTimeout(trim, lifetime);
  }
  function trim() {
    // next timer won't be able to run earlier than 10ms
    // so we'll sweep the upcoming expired entries in this run
    const now = performance.now() + 10;
    let closestExpiry = Number.MAX_SAFE_INTEGER;
    for (const [key, { expiry }] of cache.entries()) {
      if (expiry < now) {
        cache.delete(key);
      } else if (expiry < closestExpiry) {
        closestExpiry = expiry;
      }
    }
    minLifetime = closestExpiry - now;
    timer = closestExpiry < Number.MAX_SAFE_INTEGER
      ? setTimeout(trim, minLifetime)
      : 0;
  }
}
```

Reply:
Looks good to me. The only concern is about compatibility, but it should not be a problem for Chrome and Firefox.

Reply:
So do we switch or - given your preference for objects - don't?

Reply:
Hey, what about https://github.com/nicolo-ribaudo/for-own.macro
```js
import own from "for-own.macro";

for (const key in own(obj)) {
  const value = obj[key];
}
```

Reply:
I prefer `Map`. The `macro` thing is interesting though, maybe we can use it somewhere else.